### PR TITLE
GPU-AV for VK_EXT_pipeline_robustness

### PR DESCRIPTION
**Pull request by Tony-LunarG**
Adds a flags word to the start of the output buffer and a flag indicating whether to check for out of bounds buffer accesses

### CODE REVIEW

File: layers/debug_printf.cpp [hunk 1]
```diff
@@ -291,20 +291,21 @@ void snprintf_with_malloc(std::stringstream &shader_message, const DPFSubstring
...
-    //    2         Instruction Position
-    //    3         Stage Ordinal
-    //    4         Stage - specific Info Word 0
-    //    5         Stage - specific Info Word 1
-    //    6         Stage - specific Info Word 2
-    //    7         Printf Format String Id
-    //    8         Printf Values Word 0 (optional)
-    //    9         Printf Values Word 1 (optional)
-    uint32_t expect = debug_output_buffer[0];
+    //    0         Unused
```
Comment (jeremyg-lunarg) [inline layers/debug_printf.cpp:294]: should this say "must be 0"?

File: layers/debug_printf.cpp [hunk 2]
```diff
@@ -291,20 +291,21 @@ void snprintf_with_malloc(std::stringstream &shader_message, const DPFSubstring
...
+    //    6         Stage - specific Info Word 1
+    //    7         Stage - specific Info Word 2
+    //    8         Printf Format String Id
+    //    9         Printf Values Word 0 (optional)
+    //    10         Printf Values Word 1 (optional)
+    uint32_t expect = debug_output_buffer[1];
     if (!expect) return;
 
-    uint32_t index = 1;
+    uint32_t index = 2;
```
Comment (jeremyg-lunarg) [inline layers/debug_printf.cpp:308]: Is 2 kDebugOutputDataOffset from SPIRV-Tools/include/spirv-tools/instrument.hpp?

File: layers/debug_printf.cpp [hunk 3]
```diff
@@ -405,11 +406,11 @@ void DebugPrintf::AnalyzeAndGenerateMessages(VkCommandBuffer command_buffer, VkQ
         }
         index += debug_record->size;
     }
-    if ((index - 1) != expect) {
+    if ((index - 2) != expect) {
```
Comment (jeremyg-lunarg) [inline layers/debug_printf.cpp:409]: I think this is also kDebugOutputDataOffset

File: layers/debug_printf.cpp [hunk 4]
```diff
@@ -405,11 +406,11 @@ void DebugPrintf::AnalyzeAndGenerateMessages(VkCommandBuffer command_buffer, VkQ
...
         LogWarning(device, "UNASSIGNED-DEBUG-PRINTF",
                    "WARNING - Debug Printf message was truncated, likely due to a buffer size that was too small for the message");
     }
-    memset(debug_output_buffer, 0, 4 * (debug_output_buffer[0] + 1));
+    memset(debug_output_buffer, 0, 4 * (debug_output_buffer[1] + 2));
```
Comment (jeremyg-lunarg) [inline layers/debug_printf.cpp:413]: Is this `debug_output_buffer[kDebugOutputSizeOffset] + kDebugOutputDataOffset`?

File: layers/debug_printf.cpp (hunk 1 above)
Comment (Tony-LunarG) [reply layers/debug_printf.cpp:294]: Yes

File: layers/debug_printf.cpp (hunk 3 above)
Comment (Tony-LunarG) [reply layers/debug_printf.cpp:409]: Yes

File: layers/debug_printf.cpp (hunk 4 above)
Comment (Tony-LunarG) [reply layers/debug_printf.cpp:413]: Yes

**Review by jeremyg-lunarg — APPROVED** [review]
LGTM

**Comment by Tony-LunarG** [comment]
@christophe-lunarg can you inspect / approve the new GPU-AV setting?

**Review by ncesario-lunarg — APPROVED** [review]

File: layers/pipeline_state.cpp [hunk 5]
```diff
@@ -209,6 +209,32 @@ static uint32_t GetActiveShaders(const PIPELINE_STATE::StageStateVec &stages) {
...
 
+static bool UsesPipelineRobustness(const void *pNext, const PIPELINE_STATE::StageStateVec &stages) {
+    bool result = false;
+    const auto robustness_info = LvlFindInChain<VkPipelineRobustnessCreateInfoEXT>(pNext);
+    if (!robustness_info) {
+        return false;
+    }
+    result |= (robustness_info->storageBuffers == VK_PIPELINE_ROBUSTNESS_BUFFER_BEHAVIOR_ROBUST_BUFFER_ACCESS_2_EXT) ||
+              (robustness_info->storageBuffers == VK_PIPELINE_ROBUSTNESS_BUFFER_BEHAVIOR_ROBUST_BUFFER_ACCESS_EXT);
+    result |= (robustness_info->uniformBuffers == VK_PIPELINE_ROBUSTNESS_BUFFER_BEHAVIOR_ROBUST_BUFFER_ACCESS_2_EXT) ||
```
Comment (arno-lunarg) [inline layers/pipeline_state.cpp:220]: The `result |= ` logic is duplicated

File: layers/gpu_validation.cpp [hunk 6]
```diff
@@ -962,8 +963,8 @@ void GpuAssisted::PreCallRecordCreateShaderModule(VkDevice device, const VkShade
 }
 
 // Generate the part of the message describing the violation.
-bool GenerateValidationMessage(const uint32_t *debug_record, std::string &msg, std::string &vuid_msg,
-                               const GpuAssistedBufferInfo &buf_info, GpuAssisted *gpu_assisted) {
+bool GenerateValidationMessage(const uint32_t *debug_record, std::string &msg, std::string &vuid_msg, bool &oob_access,
+                               GpuAssistedBufferInfo buf_info, GpuAssisted *gpu_assisted) {
```
Comment (arno-lunarg) [inline layers/gpu_validation.cpp:967]: Was the removal of `const &` qualifier for `buf_info` intentional?

File: tests/vklayertests_gpu.cpp [hunk 7]
```diff
@@ -533,6 +533,96 @@ TEST_F(VkGpuAssistedLayerTest, GpuValidationArrayOOBGraphicsShaders) {
...
+    VkRect2D scissor = {{0, 0}, {16, 16}};
+    vk::CmdSetScissor(m_commandBuffer->handle(), 0, 1, &scissor);
+    vk::CmdDraw(m_commandBuffer->handle(), 3, 1, 0, 0);
+    vk::CmdEndRenderPass(m_commandBuffer->handle());
+    m_commandBuffer->end();
+
+    uint32_t *data = (uint32_t *)uniform_buffer.memory().map();
+    *data = 0;
+    uniform_buffer.memory().unmap();
+    m_errorMonitor->SetDesiredFailureMsg(kWarningBit, "Descriptor index 0 access out of bounds. Descriptor size is 4 and highest byte accessed was 19");
```
Comment (arno-lunarg) [inline tests/vklayertests_gpu.cpp:615]: Question: I do not get why the highest byte is 19? From my understanding, the problematic access is `u_index.index[1]`, so I figured highest byte would be 7

File: layers/gpu_validation.cpp (hunk 6 above)
Comment (Tony-LunarG) [reply layers/gpu_validation.cpp]: No it wasn't - thanks

**Pull request merged**